gradio/processing_utils.py

```diff
@@ -273,123 +273,268 @@ def save_file_to_cache(file_path: str | Path, cache_dir: str) -> str:
-    full_temp_file_path = str(abspath(temp_dir / name))
+    full_temp_file_path = str(abspath(temp_dir / filename))
 
     if not Path(full_temp_file_path).exists():
-        with (
-            sync_client.stream("GET", url, follow_redirects=True) as response,
-            open(full_temp_file_path, "wb") as f,
-        ):
-            for redirect in response.history:
-                get_public_url(str(redirect.url))
-
-            for chunk in response.iter_raw():
-                f.write(chunk)
+        with open(full_temp_file_path, "wb") as f:
+            f.write(response.content)
 
     return full_temp_file_path
 
 
-async def async_save_url_to_cache(url: str, cache_dir: str) -> str:
-    """Downloads a file and makes a temporary file path for a copy if does not already
-    exist. Otherwise returns the path to the existing temp file. Uses async httpx."""
-    url = get_public_url(url)
+async def async_ssrf_protected_download(url: str, cache_dir: str) -> str:
+    parsed_url = urlparse(url)
+    hostname = parsed_url.hostname
+
+    response = await async_get_with_secure_transport(
+        url, trust_hostname=hostname in PUBLIC_HOSTNAME_WHITELIST
+    )
+
+    while response.is_redirect:
+        redirect_url = response.headers["Location"]
+        redirect_parsed = urlparse(redirect_url)
+
+        if not redirect_parsed.hostname:
+            redirect_url = f"{parsed_url.scheme}://{hostname}{redirect_url}"
+
+        response = await async_get_with_secure_transport(redirect_url)
+
+    if response.status_code != 200:
+        raise Exception(f"Failed to download file. Status code: {response.status_code}")
+
+    content_disposition = response.headers.get("Content-Disposition")
+    if content_disposition and "filename=" in content_disposition:
+        filename = Path(content_disposition.split("filename=")[1].strip('"')).name
+    else:
+        filename = client_utils.strip_invalid_filename_characters(Path(url).name)
 
     temp_dir = hash_url(url)
     temp_dir = Path(cache_dir) / temp_dir
     temp_dir.mkdir(exist_ok=True, parents=True)
-    name = client_utils.strip_invalid_filename_characters(Path(url).name)
-    full_temp_file_path = str(abspath(temp_dir / name))
+    full_temp_file_path = str(abspath(temp_dir / filename))
 
     if not Path(full_temp_file_path).exists():
-        async with async_client.stream("GET", url, follow_redirects=True) as response:
-            for redirect in response.history:
-                get_public_url(str(redirect.url))
-
-            async with aiofiles.open(full_temp_file_path, "wb") as f:
-                async for chunk in response.aiter_raw():
-                    await f.write(chunk)
+        async with aiofiles.open(full_temp_file_path, "wb") as f:
+            async for chunk in response.aiter_bytes():
+                await f.write(chunk)
 
     return full_temp_file_path
 
 
+save_url_to_cache = ssrf_protected_download
```

Comment:
This is for custom components, right? Might be good to leave a comment explaining
```suggestion
# Custom components created with versions of gradio < 5.0 may be using the processing_utils.save_url_to_cache method, so we alias to ssrf_protected_download to preserve backwards-compatibility
save_url_to_cache = ssrf_protected_download
```